packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -239,10 +240,16 @@ export class AssetStaging extends CoreConstruct {
    */
   private stageByBundling(bundling: BundlingOptions, skip: boolean): StagedAsset {
     if (skip) {
-      // We should have bundled, but didn't to save time. Still pretend to have a hash,
-      // but always base it on sources.
+      // We should have bundled, but didn't to save time. Still pretend to have a hash.
+      // If the asset uses OUTPUT or BUNDLE, we use a CUSTOM hash to avoid fingerprinting
+      // a potentially very large source directory. Other hash types are kept the same.
+      let hashType = this.hashType;
+      if (hashType === AssetHashType.OUTPUT || AssetHashType.BUNDLE) {
```

Comment:
That's not what that means :)

Reply:
What do you mean here exactly? This is not a Docker performance issue here. It's the fact that we are fingerprinting the wrong directory.

If someone asks for a hash based on the bundle/output you currently change it to source if bundling is skipped. You introduced this in #11008. 

Before #11008, when bundling was skipped hash calculation was changed to CUSTOM.

It's a regression because we are now fingerprinting the source arbitrally and it could be a very large directory.

I agree that there should be a discussion around deferring bundling but how can we first fix this regression?

Reply:
A note on why the `NodejsFunction` uses `OUTPUT` as hash type and not `SOURCE`: it bundles not only the user's code but also all referenced node modules so it would be impractical to fingerprint the whole node modules folder. Moreover, we don't want to fingerprint .ts/.js files that represent the CDK infra code because this would incorrectly impact the hash of an asset representing runtime code.

Reply:
No I mean, 

```ts
if (hashType === AssetHashType.OUTPUT || AssetHashType.BUNDLE) {
```

was probably intended to read:

```ts
if (hashType === AssetHashType.OUTPUT || hashType === AssetHashType.BUNDLE) {
```